Gateways: Fix race for request reply

kozlovic opened the pull request:
This addresses the following race:
- client connection creates a subscription on a reply subject
- client connection sends a request
- server sends the subscription to inbound gateway
- server sends the message to outbound gateway (those may be
  to different servers)
- receiving server sends to sub interested in request subject
- app sends reply
- its server then check for interest on the reply's subject

In interestOnly mode, there is a possibility that this server
has not received the interest on the reply subject yet and would
then drop the reply.

This PR detects above scenario and will prefix the reply subject
to identify the origin cluster if it is detected that the last
subscription from the sending connection was created less than
a second ago.
Once the destination has this prefix, the destination cluster
will always send back that message to origin cluster even if
there is no registered interest.

Signed-off-by: Ivan Kozlovic <ivan@synadia.com>

derekcollison commented on server/route.go:
```diff
@@ -1038,7 +1038,12 @@ func (c *client) sendRouteSubOrUnSubProtos(subs []*subscription, isSubProto, tra
+			if !c.flushOutbound() {
+				// Another go routine is flushing already and does not
+				// have the lock. Give it a chance to finish...
+				c.mu.Unlock()
```
Add a sched here? Its a no-op but just for clarity?

kozlovic replied:
Will remove that completely. This PR was based off another that was created long before I added this logic directly in `flushOutbound()`. So will clean that up.

kozlovic commented on server/client.go:
```diff
@@ -271,6 +271,16 @@ type readCache struct {
+	// subject. This is in order to solve req/reply race where the reply may be
+	// processed in a destination cluster before the subscription interest for
+	// that reply makes it there (due to different outbound/inbound connections).
+	lastSub       *subscription
```
I went back to our original goal which was to solve the issue of "old style" request/reply where the API creates a subscription, then sends the request. I could change those two with a slice of a structure that holds the subscription and "expiration". I would then go through the list to check for a match.

kozlovic replied:
So, should I have a list here or do we care only about the very last one?

derekcollison replied:
I think we need a list but we can discuss on a GH if need be. We allow concurrent access to most connections, so even sync request/reply could produce more than one.

kozlovic commented on server/client.go:
```diff
@@ -2374,7 +2384,9 @@ func (c *client) addSubToRouteTargets(sub *subscription) {
 
 // This processes the sublist results for a given message.
-func (c *client) processMsgResults(acc *Account, r *SublistResult, msg, subject, reply []byte, collect bool) [][]byte {
+func (c *client) processMsgResults(acc *Account, r *SublistResult, msg, subject, reply []byte,
```
Will look into that.

kozlovic replied:
But you mean processMsgResults() would still call processMsgResultsEx() with pre-set args, or do you want me to duplicate the code? (we went over this already in the original PR but you no firm decision was taken I guess)

derekcollison replied:
Yes I guess that is what I am suggesting.

kozlovic replied:
I did that but also introduced flags instead of list of booleans. I have bench'ed this and it is as fast as using explicit booleans.

kozlovic commented on server/client.go:
```diff
@@ -2374,7 +2384,9 @@ func (c *client) addSubToRouteTargets(sub *subscription) {
 // This processes the sublist results for a given message.
-func (c *client) processMsgResults(acc *Account, r *SublistResult, msg, subject, reply []byte, collect bool) [][]byte {
+func (c *client) processMsgResults(acc *Account, r *SublistResult, msg, subject, reply []byte,
+	collectQueueNames, allowGWQueuesWithoutFilter bool) [][]byte {
```
Will shorten them.

kozlovic commented on server/gateway.go:
```diff
@@ -110,8 +115,9 @@ type srvGateway struct {
-	resolver netResolver // Used to resolve host name before calling net.Dial()
-	sqbsz    int         // Max buffer size to send queue subs protocol. Used for testing.
+	resolver  netResolver   // Used to resolve host name before calling net.Dial()
+	sqbsz     int           // Max buffer size to send queue subs protocol. Used for testing.
```
Now this is the local gateway structure, which means that there is only one of such object in the server. I don't think we should worry about that. Note that these two existed before, the one I added is `recSubExp`.

kozlovic commented on server/gateway.go:
```diff
@@ -218,6 +225,19 @@ func validateGatewayOptions(o *Options) error {
 
+// Computes a hash of 4 characters for the given gateway name.
+// This will be used for routing of replies.
+func getReplyPrefixForGateway(name string) []byte {
```
Only on startup when going through the list of explicit gateways, or when processing an implicit one. So should not be that often.

kozlovic commented:
@derekcollison Made some changes. Spent a lot of time doing benchmark and making sure that it does not affect the processing when no gateway are in play. For that reason, I removed `rsubs` from the readCache because I think it did affect performance (structure went over 64 bytes). I also removed the processMsgResults()->processMsgResultsEx() because I think that it too contributed a bit to a slow down. With use of flags, the "simplest" call now passes `pmrNoFlag` at the end while others can pass `pmrCollectQueueNames` and/or `pmrTreatGatewayAsClient`.
